.github/workflows/tests.yml

```diff
@@ -0,0 +1,79 @@
+name: Tests
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+
+# Declare default permissions as read only.
+permissions: read-all
+
+jobs:
+  test_linux:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        include:
+          - name: tests-linux-0
+            shards: 3
+            shard-index: 0
+          - name: tests-linux-1
+            shards: 3
+            shard-index: 1
+          - name: tests-linux-2
+            shards: 3
+            shard-index: 2
+    steps:
+      - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9
+      - name: ${{ matrix.name }}
+        run: scripts/verify_tests_on_master.sh --shards ${{ matrix.shards }} --shard-index ${{ matrix.shard-index }}
+  skp_generator:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+    steps:
+      - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9
+      - name: skp_generator 
+        run: (cd skp_generator && ./build.sh)
+  test_windows:
+    runs-on: windows-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        include:
+          - name: tests-windows-0
+            shards: 5
+            shard-index: 0
+          - name: tests-windows-1
+            shards: 5
+            shard-index: 1
+          - name: tests-windows-2
+            shards: 5
+            shard-index: 2
+          - name: tests-windows-3
+            shards: 5
+            shard-index: 3
+          - name: tests-windows-4
+            shards: 5
+            shard-index: 4
+    steps:
+      - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9
+      - name: ${{ matrix.name }}
+        run: scripts\verify_tests_on_master.bat --shards ${{ matrix.shards }} --shard-index ${{ matrix.shard-index }}
+  test_mac:
+    runs-on: macos-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        include:
+          - name: tests-macos-0
+            shards: 2 
+            shard-index: 0
+          - name: tests-macos-1
+            shards: 2
+            shard-index: 1
```

Comment:
i worry that this method of sharding is going to mean we end up missing shards one day and not noticing (like if someone just increases the "shards" number but doesn't add a new "shard-index" line). Is there any way to guarantee that all the shards are present?

Reply:
From a quick look to the docs I didn't find anything that could be used to calculate indexes programmatically before actions read the file and triggers the builds.